### ISSUE TITLE
[DISCUSSION] LineageOS and CWA: how to use?

### ISSUE THREAD

**Issue by ethpi** [corona-warn-app/cwa-app-android#741]
## How can I use the corona-warn-app with LineageOS? Ideas? Setting improvements?

I tried following setting:

* flash fresh build of LineageOS 7.1.2 an a Galaxy S3 (i9300) 
* flash suaddon
* flash  gapps pico
* install afwall+
* disconnect the play services and playstore from internet with afwall+
* do not activate a google account (perhaps freeze play store with titaniumbackup?)
* install corona-warn-app without using the playstore of this device (maybe the plays store of another device)
* connect the corona-warn-app to internet with afwall+

It works. I use this device as my coronofon with nothing else than the corona-warn-app.  Unfortunately I have to take a second device with me.

Does this settings makes sense? I hope, that the play service can't send any data to google, because it's blocked by afwall+. But I'm not very familar with the gapps. 
Any Ideas to improve the setting?

**Comment by cypheron** [comment]
@no-go You don't have to do anything. These instructions are for users who run LineageOS. I'm also excited about CoraLibre, but we have to understand that it is a community project which doesn't have a million Euro budget like CWA. Last thing I heard is that they need someone proficient with the Blueetooth stack, so I hope more people will join the project and make some real progress.
@ethpi: Thanks for the instructions!! I was looking into a way to make CWA work on LineageOS (without having to send data to Google), let's see if this works!

**Comment by no-go** [comment]
hahaha, the CWA using google closed source Tracking API and APK download discussions are really trash bags.

I am running LinageOS - so why I do not to have to do anything, if I want a really open Corona Warning App? Additionally installing and running LinageOS is possible without rooting the phone. @cypheron Setzen, sechs!

**Comment by cypheron** [comment]
@no-go Thanks for your constructive replies, they are very valuable to this discussion.
There are two kinds of people: Those who come up with ideas like @ethpi and new projects like @theScrabi, and others who only complain and demand (@no-go).

**Comment by no-go** [comment]
I didn't mean to sound so negative. The workaround is probably an interesting solution for some people who want to use gapps with LineageOS (for CWA they have to) and still want to make their system more safe. Peace!

**Comment by MilesTeg97** [comment]
> * I do not want (or I am not able) to root my phone to install a firewall app
> * I do not want to install any kind of gapps

For Everyone's Information:  relevant issue for this is #75 . 

This discussion here is obviously only about a LineageOS+gapps solution.

**Comment by treysis** [comment]
I am surprised the S3 supports BLE. It's not listed in the specs.

**Comment by ethpi** [comment]
> I am surprised the S3 supports BLE. It's not listed in the specs.

Oh, sounds bad. The CWA doesn't reports any error on my S3. I thought, CWA works correctly, because the status of "Risiko-Errmittlung" is:
internet connected / active BT : "Aktiv"
internet disconnected, active BT: "Eingeschränkt"
internet disconnected, deactivated BT: "Eingeschränkt"

How can I test, if the S3 works properly? Is there any suitable logcat command?

**Comment by MilesTeg97** [comment]
> > I am surprised the S3 supports BLE. It's not listed in the specs.
> 
> Oh, sounds bad. The CWA doesn't reports any error on my S3. I thought, CWA works correctly, because the status of "Risiko-Errmittlung" is:
> internet connected / active BT : "Aktiv"
> internet disconnected, active BT: "Eingeschränkt"
> internet disconnected, deactivated BT: "Eingeschränkt"
> 
> How can I test, if the S3 works properly? Is there any suitable logcat command?

see #668 for a method to test. Active beacon transmission is probably not working (according to this list: https://altbeacon.github.io/android-beacon-library/beacon-transmitter-devices.html see "i9300")

**Comment by treysis** [comment]
True. Beacon transmission is one thing, but I am surprised the device supports BLE even for receiving. Seems like most BT 4.0 chips come with BLE, it's just often disabled in firmware.

**Comment by doak** [comment]
I tried a similar but simplet setup with an i9305 (Galaxy SIII LTE) which don't work right now, although it looked very promising the first days:
* *pico Gapps*
* "plain" *Lineage* v14.1 (Android v7.1)
* Internet access only from time to time (via wifi) 
* No firewall, no privacy rules.

I don't know if there are any issues with bluetooth etc, but it seems even "Exposure checks" (`Settings | Google | COVID-19 Exposure Notifications | Expsoure checks`) are not working like expected: All entries show `Number of keys = 0` and `Number of matches = 0`.

Does anybody using *Gapps* see here "sane" numbers?

![Screenshot_20200709-003419](https://user-images.githubusercontent.com/2485924/86977439-e593de00-c17c-11ea-8724-06f80527c355.png)
![Screenshot_20200709-003408](https://user-images.githubusercontent.com/2485924/86977440-e62c7480-c17c-11ea-9b68-a1053034958f.png)

**Comment by ethpi** [comment]
On my i9300 it's the same. Every check shows zero "number of keys" an zero "number of matches". I checked this numbers on the phone of my son. He uses a "normal" stock ROM with the full Googleworld. I can find only the same zeros. So I hope the CWA works correctly on my i9300, because it show the same manner as on a stock ROM.

**Comment by nidico** [comment]
Same behaviour here on a Lineage 17.1 phone with gapps pico.

But the same also applies to a "normal" Samsung Android 7.0 tablet.

**Comment by doak** [comment]
> So I hope the CWA works correctly on my i9300, because it shows the same manner as on a stock ROM.
> [...]
> But the same also applies to a "normal" Trekstor Android 7.0 tablet.

Confusing. Is there **any other way** to find out if it works reliable?

That seems also tobe one the most mentioned issues by comments in Google Play Store. Just relying on a green box saying "everything is working" seems to be too less for people with some technical background and not-off-the-shelf SW.
Usual systems are quite too complex to just trust a flag without a way to make it plausible. This also holds here.
Having said this, I think the developers made a good job in showing only relevant information and making the UI as simple as possible in general.

Searching the FAQ, I have found [this](https://www.coronawarn.app/en/faq/#keys0) which seems to indicate its just a display error ...

**Comment by nidico** [comment]
>  Confusing. Is there **any other way** to find out if it works reliable?

You can try https://github.com/marq24/UUID0xFD6FTracer/ - it tracks the amount of nearby devices sending out appropriate tokens.

>  Usual systems are quite too complex to just trust a flag without a way to make it plausible. This also holds here.

I agree.

**Comment by doak** [comment]
> You can try https://github.com/marq24/UUID0xFD6FTracer/ - it tracks the amount of nearby devices sending out appropriate tokens.

I have tried and it works, needed some minutes to find a beacon (another test device). (To be honest, I even don't know how often these are sended ;)

**Comment by treysis** [comment]
Approx. every 250 ms iirc.

**Comment by doak** [comment]
> I have tried and it works, needed some minutes to find a beacon (another test device).

Not sure, but it seems I must take that back. It also finds another device even if Exposure API is turned of in Google settings :(

**Comment by doak** [comment]
I second the reading (and statements of @MilesTeg97 and @treysis), that the i9305 does **not** support beacon advertising. I tried several BLE scanning apps without success. The most reasonable were *nRF Connect* and *Locate Beacon*, but I also tried *UUID 0xFD6F Tracer* and *GPS logger II* and some more.
I also scanned for beacons with my Linux computer using the scan mode of `bluetoothctl` and `btmon` to monitor ([described here](https://ukbaz.github.io/howto/beacon_scan_cmd_line.html)). I caught several 'LE Advertising reports' from other devices (neighbourhood), but not a single from my devices (all i9305).

Sadly the i3905 (*Samsung Galaxy SIII LTE*) seems not suitable for this to work, at least it can't send. I am not sure about receiving capabilities, but there is some hope, though, since *nRF Connect* lists some beacons raw data, which were also printed by `btmon`.

**Comment by doak** [comment]
Good news: I got it working on i9305 with [some "unofficial" v16.0 Lineage build](https://forum.xda-developers.com/galaxy-s3/development-i9305/rom-lineageos-16-0-i9305-t4035331). *Peripheral mode*
 tested with *nRF Connect* which received both, a custom one sent manually and beacons from CWA.

**Please be aware that I neither know the source of the image, nor do I trust him/her.**


![Screenshot_20200710-024654_nRF_Connect](https://user-images.githubusercontent.com/2485924/87105956-90c79480-c25c-11ea-91aa-9206194878d8.png)
